### PR TITLE
Use new DlBlendMode Object

### PR DESCRIPTION
**Pull request by JsouLiang**
Use new DlBlendMode enum to change SkBlendMode

### CODE REVIEW

**Comment by JsouLiang** [comment]
cc @flar @chinmaygarde

**Comment by chinmaygarde** [comment]
I think the patch looks fine but updating the non-opaque types at this time is not the most pressing priority.

**Comment by flar** [comment]
As Chinmay says, the enums aren't the highest priority, but we should convert them as we encounter them in the larger objects. This potentially should have been done when I converted DlColorFilter.

In my most recent commit for my DlColorSource PR I converted DlTileMode as it was used by that object. Take a look at the way I did it with the conversion functions being inlined rather than declared as an actual function (the cast is a NOP, so it shouldn't force a function call).

PR: https://github.com/flutter/engine/pull/31981
File: https://github.com/flutter/engine/blob/614d9a5967c60736c2cb1e2342d113750fa883a0/display_list/display_list_tile_mode.h

**Comment by chinmaygarde** [comment]
Right. My comment about the priority was more towards convincing @JsouLiang that if they want to help out in tacking the items in https://github.com/flutter/flutter/issues/95434#issuecomment-999942639 (which I would be super appreciative of), perhaps the right ones to tackle first would be the opaque and semi-opaque ones. But the direction here especially after @flar s last comment is sound.

**Comment by flar** [comment]
The DlColorSource work has landed, which includes the enum rework for DlTileMode/SkTileMode. Take a look at that and update this for consistency (and please comment if you have suggestions on the way those are implemented). Note that I added an enum_unittests.cc to hold the (not very complicated) unit tests for the enums. They will help us discover if we ever run across a change in the Sk enums.

**Comment by JsouLiang** [comment]
@flar hey, I have change the implement of `DlBlendMode`, please give me a review, so i can join other tasks.  : )

**Comment by flar** [comment]
> @flar hey, I have change the implement of `DlBlendMode`, please give me a review, so i can join other tasks. : )

I was waiting for you to fix all of the broken checks, but I'll take a first pass look.

File: display_list/display_list_enum_unittests.cc [hunk 1]
```diff
@@ -23,5 +24,19 @@ TEST(DisplayListEnum, ToSkTileMode) {
   ASSERT_EQ(ToSk(DlTileMode::kDecal), SkTileMode::kDecal);
 }
 
+TEST(DisplayListEnum, ToDlBlendMode) {
+  ASSERT_EQ(ToDl(SkBlendMode::kSrc), DlBlendMode::kSrc);
+  ASSERT_EQ(ToDl(SkBlendMode::kLastCoeffMode), DlBlendMode::kLastCoeffMode);
+  ASSERT_EQ(ToDl(SkBlendMode::kDst), DlBlendMode::kDst);
+  ASSERT_EQ(ToDl(SkBlendMode::kLastMode), DlBlendMode::kLastMode);
+}
```
Comment (flar) [inline display_list/display_list_enum_unittests.cc:32]: I would do all of them. Cutting and pasting the values and then using a good regex replace macro should create lines to test all of them with just a little bit of typing.

File: display_list/display_list_utils.cc [hunk 2]
```diff
@@ -256,13 +258,17 @@ void DisplayListBoundsCalculator::setStrokeWidth(SkScalar width) {
...
 }
-void DisplayListBoundsCalculator::setBlendMode(SkBlendMode mode) {
+void DisplayListBoundsCalculator::setBlendMode(DlBlendMode mode) {
   blend_mode_ = mode;
 }
 void DisplayListBoundsCalculator::setBlender(sk_sp<SkBlender> blender) {
   SkPaint paint;
   paint.setBlender(std::move(blender));
-  blend_mode_ = paint.asBlendMode();
+  blend_mode_ = std::nullopt;
```
Comment (flar) [inline display_list/display_list_utils.cc:267]: Why not wait until you've tested the paint's mode and do this in an "else" case below?

**Review by flar — COMMENTED** [review]
It looks good except for a couple of minor suggestions. Also, there are changes required to pass all the submit checks, so I'll make this a comment review for now.

**Comment by flar** [comment]
Note that the impeller directory in the engine tree is imported from `github.com/flutter/impeller` so if you've made local changes to that directory, they weren't included in this push. Instead, you have to make coordinated changes to both repos.

engine/src/flutter/impeller is one of the directories synchronized by `gclient sync` (yes, it is outside of the third_party directories, but it will eventually be merged into the engine so it's not technically "3rd party")

Fork the impeller repo and make the same changes to it as you made to your local impeller copy and push those changes to your fork and make a PR for that repo. You can mark it WIP, but this all shouldn't take too long to complete.

You can test if your impeller PR works by temporarily changing the entry in your engine'd DEPS file to point to your commit in your forked impeller repo and executing `gclient sync`. (You'll need to undo the changes made to the local engine's impeller directory as it needs to be "clean" for the gclient to sync it.) If it then builds appropriately then you'll need to request a PR for your impeller changes, get that in and then update the DEPS to point to the commit hash in the official flutter/impeller repo.

Note that you still have other pieces broken in the current version of this PR before we can coordinate that dependency update. Particularly it looks like there are missing items in the licenses file?

**Comment by JsouLiang** [comment]
@flar Hey bro, I have push change to the `impeller` project, and it have approved, so next please help me merge it, and I will change the DEPS commit id. Thank you very much :)

**Review by flar — APPROVED** [review]
LGTM

**Comment by flar** [comment]
Ah, sorry, it needs a second reviewer.

**Review by zanderso — APPROVED** [review]

**Pull request merged**